packages/parse-mapping-lookup/src/ast.ts

```diff
@@ -0,0 +1,111 @@
+import debugModule from "debug";
+const debug = debugModule("parse-mapping-lookup:ast");
+
+/*
+ * Expression
+ */
+
+export interface Expression {
+  kind: "expression";
+  root: Identifier;
+  pointer: Pointer;
+}
+
+export const expression = (options: {
+  root: Identifier;
+  pointer: Pointer;
+}): Expression => {
+  const { root, pointer } = options;
+  return {
+    kind: "expression",
+    root,
+    pointer
+  };
+};
+
+/*
+ * Identifier
+ */
+
+export interface Identifier {
+  kind: "identifier";
+  name: string;
+}
+
+export const identifier = (options: { name: string }): Identifier => {
+  const { name } = options;
+  return {
+    kind: "identifier",
+    name
+  };
+};
+
+/*
+ * Pointer
+ */
+
+export interface Pointer {
+  kind: "pointer";
+  path: Step[];
+}
+
+export type Step = Access | Lookup;
+
+export const pointer = (options: { path: Step[] }): Pointer => {
+  const { path } = options;
+  return {
+    kind: "pointer",
+    path
+  };
+};
+
+/*
+ * Lookup
+ */
+
+export interface Lookup {
+  kind: "lookup";
+  property: Identifier;
+}
+
+export const lookup = (options: { property: Identifier }): Lookup => {
+  const { property } = options;
+  return {
+    kind: "lookup",
+    property
+  };
+};
+
+/*
+ * Access
+ */
+
+export interface Access {
+  kind: "access";
+  index: Literal;
+}
+
+export const access = (options: { index: Literal }): Access => {
+  const { index } = options;
+  return {
+    kind: "access",
+    index
+  };
+};
+
+/*
+ * Literal
+ */
+
+export interface Literal {
```

Comment:
So, I don't like the way literals work here -- especially the way that `0` and `"0"` are not distinguished from one another.  I'd suggest having multiple node types for literals, so that we can distinguish numeric literals and other literals from string literals.  So we could have, IDK, `text: string` for non-string literals (the actual text of the literal), and `content: string` for string literals (the content of the string, not the text of the literal), perhaps?

Reply:
Updated this to add `type` field with multiple distinct extension interfaces (`BooleanLiteral`, etc.)